source/simulator/helper_functions.cc

```diff
@@ -827,24 +860,43 @@ namespace aspect
     if (parameters.use_locally_conservative_discretization)
       AssertThrow(false, ExcNotImplemented());
 
-    // TODO: currently does not work if velocity and
-    // pressure are in the same block.
-    Assert(introspection.block_indices.velocities != introspection.block_indices.pressure,
-           ExcNotImplemented());
+    const double global_normal_velocity_integral = 0.0;
 
     // In the following we integrate the right hand side. This integral is the
     // correction term that needs to be added to the pressure right hand side.
     // (so that the integral of right hand side is set to zero).
+    if (!parameters.include_melt_transport && introspection.block_indices.velocities != introspection.block_indices.pressure)
+      {
+        const double mean       = vector.block(introspection.block_indices.pressure).mean_value();
+        const double correction = (global_normal_velocity_integral - mean * vector.block(introspection.block_indices.pressure).size()) / global_volume;
+        vector.block(introspection.block_indices.pressure).add(correction, pressure_shape_function_integrals.block(introspection.block_indices.pressure));
+      }
+    else
+      {
+        // we need to operate only on p_f not on p_c
+        const IndexSet &idxset = parameters.include_melt_transport ?
+                                 introspection.index_sets.locally_owned_fluid_pressure_dofs
+                                 :
+                                 introspection.index_sets.locally_owned_pressure_dofs;
+        double pressure_sum = 0.0;
+
+        for (unsigned int i=0; i < idxset.n_elements(); ++i)
+          {
+            types::global_dof_index idx = idxset.nth_index_in_set(i);
+            pressure_sum += vector(idx);
+          }
 
-    // We do not have to integrate over the normal velocity at the
-    // boundaries with a prescribed velocity because the constraints
-    // are already distributed to the right hand side in
-    // current_constraints.distribute.
+        const double global_pressure_sum = Utilities::MPI::sum(pressure_sum, mpi_communicator);
+        const double correction = (global_normal_velocity_integral - global_pressure_sum) / global_volume;
```

Comment:
But the global_normal_velocity_integral is set to 0.0 above, why not remove it at all? Should there be a TODO here?